framework/src/org/checkerframework/framework/type/AsSuperVisitor.java

```diff
@@ -0,0 +1,808 @@
+package org.checkerframework.framework.type;
+
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedIntersectionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedPrimitiveType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedUnionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor;
+import org.checkerframework.framework.util.AnnotatedTypes;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+/**
+ * Implements asSuper {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+ * AnnotatedTypeMirror)}
+ */
+public class AsSuperVisitor extends AbstractAtmComboVisitor<AnnotatedTypeMirror, Void> {
+
+    private final Types types;
+    private final AnnotatedTypeFactory annotatedTypeFactory;
+
+    public AsSuperVisitor(AnnotatedTypeFactory annotatedTypeFactory) {
+        this.annotatedTypeFactory = annotatedTypeFactory;
+        types = annotatedTypeFactory.types;
+    }
+
+    /**
+     * Implements asSuper {@link AnnotatedTypes#asSuper(AnnotatedTypeFactory, AnnotatedTypeMirror,
+     * AnnotatedTypeMirror)}
+     *
+     * @param type      Type from which to copy annotations
+     * @param superType a type whose erased Java type is a supertype of {@code type}'s erased Java
+     *                  type.
+     * @return {@code superType} with annotations copied from {@code type} and type variables
+     * substituted from {@code type}.
+     */
+    @SuppressWarnings("unchecked")
+    public <T extends AnnotatedTypeMirror> T asSuper(AnnotatedTypeMirror type, T superType) {
+        if (type == null || superType == null) {
+            ErrorReporter.errorAbort("AsSuperVisitor type and supertype cannot be null.");
+            return null; // dead code
+        } else if (type == superType) {
+            return (T) type.deepCopy();
```

Comment:
The pre and post conditions of asSuper are documented in AnnotatedTypes#asSuper.  I improved the Javadoc to point to that method.